server/src/main/java/org/apache/druid/catalog/model/table/BaseInputSourceDefn.java

```diff
@@ -0,0 +1,278 @@
+{
+  /**
+   * The "from-scratch" table function for this input source. The parameters
+   * are those defined by the subclass, and the apply simply turns around and
+   * asks the input source definition to do the conversion.
+   */
+  public class AdHocTableFunction extends BaseTableFunction
+  {
+    public AdHocTableFunction(List<ParameterDefn> parameters)
+    {
+      super(parameters);
+    }
+
+    @Override
+    public ExternalTableSpec apply(
+        final String fnName,
+        final Map<String, Object> args,
+        final List<ColumnSpec> columns,
+        final ObjectMapper jsonMapper
+    )
+    {
+      requireSchema(fnName, columns);
+      return convertArgsToTable(args, columns, jsonMapper);
+    }
+  }
+
+  /**
+   * The "partial" table function that starts with a catalog external table spec, then
+   * uses SQL function arguments to "complete" (i.e. fill in) the missing properties to
+   * produce a complete table which is then converted to an external table which Calcite
+   * can use.
+   * <p>
+   * The set of parameters depends on the input source and on whether or not the catalog
+   * spec provides a format.
+   */
+  public class PartialTableFunction extends BaseTableFunction
+  {
+    private final ResolvedExternalTable table;
+
+    public PartialTableFunction(final ResolvedExternalTable table, List<ParameterDefn> params)
+    {
+      super(params);
+      this.table = table;
+    }
+
+    @Override
+    public ExternalTableSpec apply(
+        final String fnName,
+        final Map<String, Object> args,
+        final List<ColumnSpec> columns,
+        final ObjectMapper jsonMapper
+    )
+    {
+      if (CollectionUtils.isNullOrEmpty(table.resolvedTable().spec().columns())) {
+        requireSchema(fnName, columns);
+      }
+      return convertCompletedTable(table, args, columns);
+    }
+  }
+
+  /**
+   * The one and only from-scratch table function for this input source. The
+   * function is defined a bind time, not construction time, since it typically
+   * needs visibility to the set of available input formats.
+   */
+  private AdHocTableFunction adHocTableFn;
+
+  /**
+   * Overridden by each subclass to return the input source class to be
+   * used for JSON conversions.
+   */
+  protected abstract Class<? extends InputSource> inputSourceClass();
+
+  @Override
+  public void bind(TableDefnRegistry registry)
+  {
+    this.adHocTableFn = defineAdHocTableFunction();
+  }
+
+  @Override
+  public void validate(ResolvedExternalTable table)
+  {
+    convertTableToSource(table);
+  }
+
+  /**
+   * Overridden by each subclass to define the parameters needed by each
+   * input source.
+   */
+  protected abstract AdHocTableFunction defineAdHocTableFunction();
+
+  @Override
+  public TableFunction adHocTableFn()
+  {
+    return adHocTableFn;
+  }
+
+  /**
+   * Define a table "from scratch" using SQL function arguments.
+   */
+  protected ExternalTableSpec convertArgsToTable(
```

Comment:
This this be moved into the `AdHocTableFunction` class, since it seems it should only be used in that case?